port/unix/omrsysinfo.c

```diff
@@ -5640,7 +5670,139 @@ readCgroupFile(struct OMRPortLibrary *portLibrary, int pid, BOOLEAN inContainer,
+			Trc_PRT_populateCgroupEntryListV1_available_subsystems(available);
+		}
+	}
+
+	return rc;
+}
+
+/**
+ * Reads /proc/<pid>/cgroup to get the name of the cgroup to which the given process
+ * belongs, then the $MOUNT_POINT/cgroupName/cgroup.controllers file to get the
+ * enabled subsystems.
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] pid process id
+ * @param[out] cgroupEntryList pointer to OMRCgroupEntry *. On successful return, *cgroupEntry
+ * points to a circular linked list. Each element of the list is populated based on the contents
+ * of /proc/<pid>/cgroup file.
+ * @param[out] availableSubsystems on successful return, contains bitwise-OR of flags of type OMR_CGROUP_SUBSYSTEMS_*
+ * indicating the subsystems available for use
+ *
+ * returns 0 on success, negative code on error
+ */
+static int32_t
+populateCgroupEntryListV2(struct OMRPortLibrary *portLibrary, int pid, OMRCgroupEntry **cgroupEntryList, uint64_t *availableSubsystems)
+{
+	char cgroupFilePath[PATH_MAX];
+	char controllerFilePath[PATH_MAX];
+	char buffer[PATH_MAX];
+	uintptr_t requiredSize = 0;
+	FILE *cgroupFile = NULL;
+	FILE *controllerFile = NULL;
+	OMRCgroupEntry *cgEntryList = NULL;
+	uint64_t available = 0;
+	int32_t rc = 0;
+
+	Assert_PRT_true(NULL != cgroupEntryList);
+
+	requiredSize = portLibrary->str_printf(portLibrary, NULL, (uint32_t)-1, "/proc/%d/cgroup", pid);
+	Assert_PRT_true(requiredSize <= PATH_MAX);
+	portLibrary->str_printf(portLibrary, cgroupFilePath, sizeof(cgroupFilePath), "/proc/%d/cgroup", pid);
+
+	cgroupFile = fopen(cgroupFilePath, "r");
+	if (NULL == cgroupFile) {
+		int32_t osErrCode = errno;
+		Trc_PRT_populateCgroupEntryListV2_fopen_failed(cgroupFilePath, osErrCode);
+		rc = portLibrary->error_set_last_error(portLibrary, osErrCode, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_FOPEN_FAILED);
+		goto _end;
+	}
+
+	if (NULL == fgets(buffer, PATH_MAX, cgroupFile)) {
+		if (0 != ferror(cgroupFile)) {
+			int32_t osErrCode = errno;
+			Trc_PRT_populateCgroupEntryListV2_fgets_failed(cgroupFilePath, osErrCode);
+			rc = portLibrary->error_set_last_error_with_message_format(portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED, "fgets failed to read %s file stream with errno=%d", cgroupFilePath, osErrCode);
+			goto _end;
+		}
+	} else {
+		/* For v2, cgroup file contains one line of the format 0::<cgroup> */
+		char cgroup[PATH_MAX];
+		/* This array should be large enough to read names of all subsystems. 1024 should be enough based on current supported subsystems. */
+		char subsystems[PATH_MAX];
+		char *cursor = NULL;
+		char *separator = NULL;
+
+		rc = sscanf(buffer, PROC_PID_CGROUPV2_ENTRY_FORMAT, cgroup);
+
+		if (1 == rc) {
+			/* Controller file consists of single space-delimited line of controllers/subsystems */
+			requiredSize = portLibrary->str_printf(portLibrary, NULL, (uint32_t)-1, "%s%s/cgroup.controllers", OMR_CGROUP_MOUNT_POINT, cgroup);
```

Comment:
The first is the root cgroup, controllers file here determines the available controllers for all cgroups. The second is the controllers available for a certain cgroup which is a subset of the root cgroup controllers. I don't think `cgroup` needs to be renamed since the path is essentially the name